onnx/onnx.proto

```diff
@@ -212,6 +237,44 @@ message TensorProto {
 message SparseTensorProto {
   // The dimensions in the tensor.
   repeated int64 dims = 1;
+  // This field MUST be present this version of the IR.
   optional TensorProto indices = 2;
+  // This field MUST be present this version of the IR.  
   optional TensorProto values = 3;
 }
+
+// Define the types.
+message TypeProto {
+  message TensorTypeProto {
+  // This field MUST NOT have the value of UNDEFINED
+  // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  message SparseTensorTypeProto {
+  // This field MUST NOT have the value of UNDEFINED
+  // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  oneof value {
+    // The type of a tensor.
+    TensorTypeProto tensor_type = 1;
+
+    // The type of a sparse tensor.
+    SparseTensorTypeProto sparse_tensor_type = 2;
+  }
+}
+
+// Defines a tensor shape. A dimension can be either an integer value
+// or a symbolic variable. A symbolic variable represents an unknown
+// dimension.
+message TensorShapeProto {
+  message Dimension {
+    oneof value {
+      int64 dim_value = 1;
+      string dim_param = 2;   // namespace Shape
+    };
+  };
+  repeated Dimension dim = 1;
+}
```

Comment:
Should we also support arbitrary list of dimensions? E.g. we can have "named range of dimensions". One option to specify it would be to put `string dim_list` above that can cover 0 or more arbitrary dimensions preserved "as is".

For example `Gather` operator would be taking shapes `[N, ...dims...]` and `[K]` and returning `[K, ...dims...]`.

Reply:
I've hit the need for something like this as well. If I understand correctly, you are saying we'd effectively add a rest arg/variatic arg and allow it to be named, again, to support correlation?

Reply:
Yes, adding a "variadic placeholder" that can capture zero or more dimensions

Reply:
TF has a more general `Gather` where the axis to gather is an input. Do we want to handle it?

Reply:
It's actually a perfect example why static shape inference might be limiting and providing C++ inference function is beneficial. E.g. that's exactly what TF does in this case: https://github.com/tensorflow/tensorflow/blob/3c0bcf8442a107b635cca64440d05c390a35d177/tensorflow/core/ops/array_ops.cc#L1396

I think for now we can avoid "vararg" expressions. For inputs/outputs specification even simple symbolic names like in this PR should be sufficient.